api/agent/agent.go

```diff
@@ -855,20 +825,32 @@ func (a *agent) runHot(ctx context.Context, caller slotCaller, call *call, tok R
 		}
 	}()
 
+	// Monitor initialization and evictability.
 	go func() {
-		isInitialized, err := a.waitInitialize(ctx, caller, evictor, udsWait)
-		if err != nil {
-			tryQueueErr(err, errQueue)
-		}
-		if isInitialized {
-			close(initialized)
-		} else {
-			cancel()
+		for {
+			select {
+			case err := <-udsWait:
+				if tryQueueErr(err, errQueue) != nil {
+					cancel()
+				}
+				return
+			case <-ctx.Done(): // container shutdown
+				return
+			case <-a.shutWg.Closer(): // agent shutdown
+				cancel()
+				return
+			case <-caller.done: // original caller disconnected or serviced by another container?
+				evictor.SetEvictable(true)
+				caller.done = nil // block 'caller.done' after this point
+			case <-evictor.C: // eviction
+				cancel()
```

Comment:
am curious about https://github.com/fnproject/fn/pull/1288/commits/af5236241d131d58d3e20c0b2478353a90f5ee79#r234048169 for this and the udsWait too - i'm not 100% sure, it may be okay, worth thinking about maybe, i'll think too...